initial backend implementation

berndverst opened the pull request.

berndverst committed:
Initial implementation of DTMB backend

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Filter tasks on backend

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Filter tasks on backend continued

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
All tests pass

berndverst committed:
dtmb backend implementation continued

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
WIP

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
WIP: Does not compile

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
start caching of history

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
lots of progress

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
more backend stuff

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Add updated protos

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
More backend implementation

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
implement remaining methods

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Add some cache evictions

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
More backend stuff

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
more things

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
more things'

Signed-off-by: Bernd Verst <github@bernd.dev>

ItalyPaleAle commented on backend/azure/dtmb/internal/backend/v1/client.proto:
```diff
@@ -0,0 +1,115 @@
+syntax = "proto3";
```
I assume these protos are here temporarily right? Eventually, they'll be in their own repos?

berndverst replied:
Yes - I am waiting to hear from you where we'll have the protos :)

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/connectworker.go:
```diff
@@ -0,0 +1,248 @@
+
+	// Establish the ConnectWorker stream
+	streamCtx := metadata.AppendToOutgoingContext(ctx,
+		"taskhub", taskHub,
```
Nit: use a constant for key

berndverst replied:
Yeah, like I said this PR is nowhere near final - my TODO is to make a lot of constants configurable where the backend is initiated.

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/connectworker.go:
```diff
@@ -0,0 +1,248 @@
+			return
+		}
+
+		log.Printf("[%s] Received configuration message: %v", testID, wc)
```
Remove logs

berndverst replied:
Later! Right now I wanted just a high level review from you or what the specific calls to/from the backend

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/connectworker.go:
```diff
@@ -0,0 +1,248 @@
+	go func() {
+		msg, err := stream.Recv()
+		if err != nil {
+			configReceived <- err
```
```suggestion
			configReceived <- fmt.Errorf("error receiving initial message: %w", err)
```

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/connectworker.go:
```diff
@@ -0,0 +1,248 @@
+
+			case <-tick.C:
+				if debug {
+					log.Printf("[%s] Sending ping…", testID)
```
Probably remove these logs too?

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/connectworker.go:
```diff
@@ -0,0 +1,248 @@
+	return res
+}
+
+func generateTestID() (string, error) {
```
Remove this from my test code :)

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/datastructures.go:
```diff
@@ -0,0 +1,144 @@
+	q.items = append(q.items, item)
+}
+
+func (q *SyncQueue[T]) EnqueueBatch(items []*T) {
```
Should this use the `...` signature?

berndverst replied:
Yeah - one method instead of two similar ones. Good idea.

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/datastructures.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+func (q *SyncQueue[T]) ReadAllWithoutDequeue() []*T {
+	q.lock.Lock()
```
Could we use a RW MUtex for these operations? So that read operations would not block each other? Unless there's only 1 consumer for the queue

berndverst replied:
I had two separate lock types previously but then changed it. Yeah that's fine.

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/datastructures.go:
```diff
@@ -0,0 +1,144 @@
+	return q.items
+}
+
+func (q *SyncQueue[T]) ReadLastWithoutDequeue() *T {
```
Nit: I think this operation is normally called "peek", so "PeekLast"

berndverst replied:
I'm aware of Peek - but I've never seen PeekLast so wasn't sure on naming. PeekLast is fine by me though.

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/datastructures.go:
```diff
@@ -0,0 +1,144 @@
+func (q *SyncQueue[T]) ReadAllWithoutDequeue() []*T {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	return q.items
```
I would recommend returning a copy of `q.items` otherwise the caller could modify the data inside the queue.
https://pkg.go.dev/slices#Clone

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/datastructures.go:
```diff
@@ -0,0 +1,144 @@
+
+	capacity int
+	cache    map[string]*list.Element
+	list     *list.List
```
What's the reason for using this data type?

berndverst replied:
LRU implementation! We can easily move a list element (which is a particular orchestration history stack) to the front of the list (meaning used most recently). If we exceed our capacity (list length) we can remove the last element in the list.

There is some synergy between the list Elements and list length.

ItalyPaleAle replied:
Oh great!

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/datastructures.go:
```diff
@@ -0,0 +1,144 @@
+		}
+		element.Value.(*orchestrationHistoryItem).historyEvents.EnqueueBatch(events)
+
+		//= orchestrationHistoryItem{orchestrationId: key, value: value}
```
`return` here so you can avoid an `else` block and make code more readable!

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/typeconversion.go:
```diff
@@ -0,0 +1,315 @@
+	}
+}
+
+func ConvertEvent(event *dtmbprotos.Event) (*protos.HistoryEvent, error) {
```
NGL this kinda hurts the eyes :)

Please consider splitting it into a method for each history event... such as `convertExecutionCompleted`, etc.

berndverst replied:
You did this to me :D 

I'll see what I can do. No promises on this one.

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/typeconversion.go:
```diff
@@ -0,0 +1,315 @@
+	switch typedEvent := event.GetEventType().(type) {
+	case *dtmbprotos.Event_ExecutionStarted:
+		return &protos.HistoryEvent{
+			EventId:   int32(event.GetSequenceNumber()), // is this correct?
```
Yes

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/typeconversion.go:
```diff
@@ -0,0 +1,315 @@
+				ExecutionStarted: &protos.ExecutionStartedEvent{
+					Name:    typedEvent.ExecutionStarted.GetName(),
+					Version: &wrapperspb.StringValue{Value: typedEvent.ExecutionStarted.GetVersion()},
+					Input:   &wrapperspb.StringValue{Value: string(typedEvent.ExecutionStarted.GetInput())}, // what if this is not string
```
DTFx SDKs support strings only, although DTMB supports generic binary data. As long as you use DTFx only, it will work

berndverst replied:
@cgillum should the protos in Durable Task be updates to support either bytes or string for Input / Output? Right now the protos can only accept strings and this seem problematic unless you want me to base64 encode all input, base64 decode the result etc

cgillum replied:
No, all SDKs and engines exclusively use strings for data and won’t be able to benefit from bytes. Just convert them to/from UTF-8 bytes when calling Ale’s APIs. That or we should change the APIs to accept strings.

ItalyPaleAle replied:
Converting from a UTF8 string to a byte slice is always possible and safe. But storing binary (i.e. non-UTF8) data in a string requires encoding to formats like Base64. Unless we have a reason for saying otherwise, I'd prefer the backend to work with byte slices since they are opaque payloads anyways. Clients can interpret the byte slices as UTF8 strings if they wish.

berndverst replied:
@ItalyPaleAle @cgillum from the Go documentation:
> It’s important to state right up front that a string holds arbitrary bytes. It is not required to hold Unicode text, UTF-8 text, or any other predefined format. As far as the content of a string is concerned, it is exactly equivalent to a slice of bytes.

So I suppose I can just treat the arbitrary bytes as string... whether that's actually UTF-8 or ASCII, I will not know.

Basically I can treat string as synonym for a byte slice.

In this case I will NOT elect to do any encoding - otherwise we'd be returning encoded data for this backend which we wouldn't be returning in other backends.

ItalyPaleAle replied:
Yes the Go documentation is correct. But in this case, if we know that all current SDKs only use strings, then you can be sure that the data that is in that `byte[]` is a UTF8-encoded string. So the conversion `byte[] -> string` is safe.

ASCII standard (0-127) is a subset of UTF8, so there's no issue there.

berndverst replied:
Yes agreed.

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/typeconversion.go:
```diff
@@ -0,0 +1,315 @@
+					},
+					ParentInstance: &protos.ParentInstanceInfo{
+						TaskScheduledId: int32(typedEvent.ExecutionStarted.GetParent().GetSequenceNumber()),
+						Name:            &wrapperspb.StringValue{Value: typedEvent.ExecutionStarted.GetParent().GetName()},
```
Nit: consider creating a helper method `toStringValue`...

..which accepts `string` and `byte[]` too if you want to make your life easier!

berndverst replied:
@cgillum @ItalyPaleAle should we do a base64 conversion if the type is byte[] ?

ItalyPaleAle replied:
Given what Chris said about currently us using strings only, I think supporting strings only is fine for now.

berndverst replied:
Yeah sounds good

ItalyPaleAle commented on backend/azure/dtmb/internal/utils/typeconversion.go:
```diff
@@ -0,0 +1,315 @@
+			Timestamp: event.GetTimestamp(),
+			EventType: &protos.HistoryEvent_ExecutionTerminated{
+				ExecutionTerminated: &protos.ExecutionTerminatedEvent{
+					Input:   &wrapperspb.StringValue{Value: string(typedEvent.ExecutionTerminated.GetReason())}, // what if this is not string
```
Same as the other comment (and for the other ones below too)

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+	be.orchestrationHistoryCache = utils.NewOrchestrationHistoryCache(nil) // TODO: make capacity configurable
+
+	ctx := context.Background()
+	connCtx, connCancel := context.WithTimeout(ctx, 15*time.Second) // TODO: make this a configurable timeout
```
Should probably change the signature of `NewDTMB` to accept a context from the user, so the user can set a timeout

cgillum replied:
I think of context as being associated with a request. Should the timeout instead by a property of `DTMBOptions`?

ItalyPaleAle replied:
There's also a context associated with the app's lifecycle. It's common pattern to have a context started when the app starts, and cancel it when the app is shutting down gracefully

berndverst replied:
@cgillum that's what my comment is saying - make the timeout configurable in DTMBOptions.

But anyways, I was not looking for a review on those details yet - I will be refactoring some of the options anyway. I am worried about the data mapping for now.

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+//
+// If the task hub for this backend already exists, an error of type [ErrTaskHubExists] is returned.
+func (d *dtmb) CreateTaskHub(ctx context.Context) error {
+	_, err := d.clientClient.Metadata(ctx, &dtmbprotos.MetadataRequest{})
```
Add a comment explaining that on DTMB the Task Hub must be created on the service, and here we just check if it exists

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+//
+// If the task hub for this backend doesn't exist, an error of type [ErrTaskHubNotFound] is returned.
+func (d *dtmb) DeleteTaskHub(context.Context) error {
+	return fmt.Errorf("the TaskHub cannot be deleted via the API. Please delete the TaskHub manually")
```
```suggestion
	return errors.New("the TaskHub cannot be deleted using this SDK; please perform this operation on the service")
```

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+func (d *dtmb) Start(ctx context.Context, orchestrators *[]string, activities *[]string) error {
+	// TODO: is the context provided a background context?
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	d.workerCancelFunc = cancel
+
+	err := d.connectWorker(ctxWithCancel, orchestrators, activities)
```
Nit:


```suggestion
	ctx, d.workerCancelFunc = context.WithCancel(ctx)
	err := d.connectWorker(ctx, orchestrators, activities)
```

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+
+// Start starts any background processing done by this backend.
+func (d *dtmb) Start(ctx context.Context, orchestrators *[]string, activities *[]string) error {
+	// TODO: is the context provided a background context?
```
At the start of this method, add a check to make sure that the object hasn't already been started, or bad things™ will happen. For exmaple use an `atomic.Bool running` and do `if !running.CompareAndSwap(false, true) { error }`

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+	// Establish the ConnectWorker stream
+	worker := d.workerClient
+	var taskHubName string = "MAKE THIS CONFIGURABLE"
+	var testID string = "MAKE THIS CONFIGURABLE"
```
This was used for my test app, but should not be here

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+				}()
+
+			case *dtmbprotos.ConnectWorkerServerMessage_ExecuteOrchestration:
+				go func() {
```
Do we need to do this in a background goroutine you think?

berndverst replied:
I'll remove for now and we can evaluate later.

ItalyPaleAle commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+				Time: executionStartedEvent.GetScheduledStartTimestamp(),
+			},
+		},
+		IdReusePolicy: &dtmbprotos.OrchestrationIDReusePolicy{ // is this correct?
```
ID reuse policy isn't implemented yet in the service

ItalyPaleAle commented on backend/backend.go:
```diff
@@ -49,7 +49,8 @@ type Backend interface {
 	// Start starts any background processing done by this backend.
-	Start(context.Context) error
+	// Backends supporting this feature can filter orchestrator and activity types server-side.
+	Start(ctx context.Context, orchestrators *[]string, activities *[]string) error
```
Why a pointer to a slice? This is unusual. Do you have to modify the slice?

cgillum replied:
I had the same question about this.

berndverst replied:
I keep forgetting nil is a value I can use for slices :) No reason other than the ability to check for nil / use nil.

ItalyPaleAle commented on backend/executor.go:
```diff
@@ -404,8 +404,8 @@ loop:
 // mustEmbedUnimplementedTaskHubSidecarServiceServer implements protos.TaskHubSidecarServiceServer
-func (grpcExecutor) mustEmbedUnimplementedTaskHubSidecarServiceServer() {
-}
+func (grpcExecutor) mustEmbedUnimplementedTaskHubSidecarServiceServer() { //nolint:unused
```
No, this is not how this method is meant to be used.

This is here because gRPC wants you to embed the unimplemented `TaskHubSidecarServiceServer` struct...

berndverst replied:
I didn't change anything here... I only added the linter comment

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+	orchestrationHistoryCache utils.OrchestrationHistoryCache
+	connectWorkerClientStream chan *dtmbprotos.ConnectWorkerClientMessage
+	clientClient              dtmbprotos.TaskHubClientClient
+	workerClient              dtmbprotos.TaskHubWorkerClient
```
Backends shouldn't be concerned about "clients" or "workers". Rather, these are concepts that belong at the upper layers, i.e. the user-facing SDKs. Not your issue, but it's making me realize that we may need to revisit the design of the gRPC API.

ItalyPaleAle replied:
The reason it's split into two is because one could have a client-only app, so the methods are separate.

However, @berndverst you don't need 2 objects. You can create something like:

```go
type TaskHubClient struct {
    dtmbprotos.TaskHubClientClient
    dtmbprotos.TaskHubWorkerClient
}
```

Then use embedding to add both "clients" to this, and you have a single object.

It's actually meant to be used this way when you have an app that needs both.

berndverst committed:
Update backend/azure/dtmb/internal/utils/connectworker.go

Co-authored-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

berndverst committed:
Update backend/azure/dtmb/dtmb.go

Co-authored-by: Alessandro (Ale) Segala <43508+ItalyPaleAle@users.noreply.github.com>

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+		OrchestrationId: executionStartedEvent.GetOrchestrationInstance().GetInstanceId(),
+		Name:            executionStartedEvent.GetName(),
+		Version:         executionStartedEvent.GetVersion().GetValue(),
+		Input:           []byte(executionStartedEvent.GetName()),
```
Looks like this should actually be `GetInput().GetValue()`?

berndverst replied:
Copy paste error :)

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+// AddNewEvent adds a new orchestration event to the specified orchestration instance.
+func (d *dtmb) AddNewOrchestrationEvent(ctx context.Context, id api.InstanceID, event *backend.HistoryEvent) error {
+	var err error
+	switch typedEvent := event.GetEventType().(type) {
```
My preference would be that the backend API doesn't need to care about what the type of the event is (and that this code doesn't need to care as well), though I need to take a closer look to understand how practical this is.

berndverst replied:
I suppose that's a comment for @ItalyPaleAle

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+			Input:           []byte(typedEvent.EventRaised.GetInput().GetValue()), // or should this be backend.MarshalHistoryEvent(event)?
+		}
+
+		_, err = d.clientClient.RaiseEvent(context.Background(), &req)
```
Any reason not to use `ctx` here?

berndverst replied:
No reason, autocomplete. Fixed.

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+		return nil, err
+	}
+
+	events := res.GetEvent()
```
Should this be named `GetEvents()`, plural?

berndverst replied:
That would require a change in the proto cc @ItalyPaleAle

ItalyPaleAle replied:
I will change that in the next iteration.

In the meanwhile, feel free to change the `proto` in this PR. Even if the name is different, they are compatible on the wire, so it will still work.

berndverst replied:
I'll wait instead.

ItalyPaleAle replied:
Ok. I've made the change in the current open PR.

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+	events := res.GetEvent()
+
+	// cache the newly received events
+	d.orchestrationHistoryCache.AddHistoryEventsForOrchestrationID(orchestrationID, events)
```
I haven't yet gotten to a chance to look at the implementation yet, but lazy me is curious: do we need to be worried about synchronization, if multiple goroutines are fetching (and possibly updating) the cached history at the same time?

berndverst replied:
The cache implementation has read/write locks, so no worries there

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+
+// GetOrchestrationMetadata gets the metadata associated with the given orchestration instance ID.
+//
+// Returns [api.ErrInstanceNotFound] if the orchestration instance doesn't exist.
```
I didn't see this error value being used, so just curious: how are we handling the case where the orchestration doesn't exist? Is there some value is `resp` that would indicate this?

berndverst replied:
It's not totally clear what errors are returned in these cases. In general, returned errors from the API are something we will have to cover in a second pass. I haven't seen any details on what errors look like.

ItalyPaleAle replied:
If the orchestration doesn't exist the gRPC request will fail with a gRPC error (`err` will contain a gRPC error). You can get the status code from that error and it will be a "NotFound". More details are in the error description.

ItalyPaleAle replied:
In general, errors returned by the gRPC APIs follow the typical pattern for gRPC errors. For Go, this is the key method: https://pkg.go.dev/google.golang.org/grpc/status#Code

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+				Version:         version,
+				CompletionToken: completionToken,
+				CustomStatus:    item.State.CustomStatus.GetValue(),
+				Actions:         nil, // // what do I put here?
```
The backend isn't normally supposed to be concerned about actions. We may need to revisit this part of the gRPC API.

berndverst replied:
FYI @ItalyPaleAle

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+		},
+	}
+	// this is a terminal event, so we can evict the cache
+	d.orchestrationHistoryCache.EvictCacheForOrchestrationID(string(item.InstanceID))
```
What do you mean by "terminal"? This API completes the work item, but not the orchestration. Not sure if that was already understood.

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+
+	// this can't be the right way to provide the input, but how should it be done?
+
+	genericInputWrapper := protos.HistoryEvent{
```
The code that executes the activity work item is expecting a `TaskScheduledEvent`.

berndverst replied:
Ok great - I had that before but then wasn't sure.

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+				OrchestrationId: string(item.InstanceID),
+				Name:            activityName,    // what do I put here?
+				CompletionToken: completionToken, // what do I put here?
+				Result:          bytes,           // is this correct?
```
It looks like you're sending a serialized history event? I assume that this is not what Alessandro's implementation is expecting. I think it should be the serialized data response from the activity function, which corresponds to `TaskCompletedEvent.GetResult().GetValue()`.

berndverst replied:
Yeah that's also what I had before - but I kept changing things back and forth because of the type mismatch -- byte[] vs string. And the SQLite backend uses a serialized history event.

I'll change it back.

cgillum commented on backend/azure/dtmb/dtmb.go:
```diff
@@ -0,0 +1,531 @@
+		Request: &dtmbprotos.PurgeOrchestrationRequest_OrchestrationId{
+			OrchestrationId: string(id),
+		},
+		NonRecursive: false, // TODO: Should this be globally configurable?
```
Ideally, we would always make this non-configurable. We can consider it later, though - I don't expect anyone will want to enable non-recursive.

berndverst replied:
Sounds good.

cgillum reviewed (commented):
I've reviewed most everything at a high level except for utils, which I assume you and Alessandro have a good handle on. Overall LGTM so far, though I have a few follow-ups to look at on the API side, as I noted.

berndverst committed:
fixup

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Address code review comments

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Another round of updates

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
fix interface

Signed-off-by: Bernd Verst <github@bernd.dev>

dependabot[bot] committed:
Bump golang.org/x/net from 0.17.0 to 0.23.0 (#68)

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.17.0 to 0.23.0.
- [Commits](https://github.com/golang/net/compare/v0.17.0...v0.23.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump golang.org/x/net from 0.17.0 to 0.23.0 (#68)

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.17.0 to 0.23.0.
- [Commits](https://github.com/golang/net/compare/v0.17.0...v0.23.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

berndverst committed:
Actions

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
some bug fixes

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
significant refactor

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
BUGFIXES

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Update code for latest proto changes

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Update protos again

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Add Azure Auth support

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Add DisableAuth option

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Fix timer problem

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Temporarily add sample apps

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Disable timers for now

Signed-off-by: Bernd Verst <github@bernd.dev>

ItalyPaleAle committed:
Make deadlines work

cgillum committed:
Switch from MIT to Apache 2.0 License (#66)

Switching to the Apache 2.0 License is being done for two reasons:
1. The original Durable Task Framework is also an Apache 2.0 License project.
2. This project is a key dependency for Dapr, a CNCF incubating project, and moving this dependency to Apache 2.0 will simplify the graduation process for Dapr (of which Microsoft is also a key contributor).

berndverst committed:
Update protos

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
refactoring

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Use Azure CLI Credential in Sample

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
linter

Signed-off-by: Bernd Verst <github@bernd.dev>

mikeee committed:
ci: add go1.22 to the matrix and remove go1.19 runs

ItalyPaleAle committed:
Merge pull request #71 from mikeee/patch-1

ci: add go1.22 to the matrix and remove go1.19 runs

berndverst committed:
Merge remote-tracking branch 'upstream/main' into dtmbbackend

berndverst committed:
update go.mod

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
update proto again

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Several bugfixes and auth settings

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
More sample updates

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
handle default port

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
fix insecure connection

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Update sequence sample

Signed-off-by: Bernd Verst <github@bernd.dev>

ItalyPaleAle committed:
💄

ItalyPaleAle committed:
Updated required protos

ItalyPaleAle committed:
Add support for reconnecting

ItalyPaleAle committed:
Add retryClientMsgChan

berndverst committed:
Refactor DTS Client with options pattern

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
update samples

Signed-off-by: Bernd Verst <github@bernd.dev>

ItalyPaleAle committed:
Merge branch 'dtmbbackend' of https://github.com/berndverst/durabletask-go into dtmbbackend-patch1

berndverst committed:
Merge pull request #2 from ItalyPaleAle/dtmbbackend-patch1

[dtmbbackend] Add support for reconnection in ConnectWorker

berndverst committed:
implement ID reuse policy

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
more cleanup

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Update go.mod, especially protobuf

Signed-off-by: Bernd Verst <github@bernd.dev>

berndverst committed:
Some cleanup

Signed-off-by: Bernd Verst <github@bernd.dev>
